File: domain/src/main/java/siarhei/luskanau/iot/doorbell/ui/BaseRecyclerArrayAdapter.java
package siarhei.luskanau.iot.doorbell.ui;

import android.support.v7.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class BaseRecyclerArrayAdapter<T, V extends RecyclerView.ViewHolder> extends BaseRecyclerAdapter<V> {

    private final List<T> mData = new ArrayList<>();

    public BaseRecyclerArrayAdapter() {
    }

    public BaseRecyclerArrayAdapter(final OnItemClickListener<V> listener) {
        super(listener);
    }

    public void addItem(final T item) {
        mData.add(item);
        notifyItemInserted(mData.size() - 1);
    }

    public void addItem(final int position, final T item) {
        mData.add(position, item);
        notifyItemInserted(position);
    }

    public void addItems(final Collection<? extends T> items) {
        if (items != null) {
            final int oldSize = getItemCount();
            mData.addAll(items);
            notifyItemRangeInserted(oldSize, getItemCount() - oldSize);
        }
    }

    public void addItems(final int position, final Collection<? extends T> items) {
        if (items != null) {
            mData.addAll(position, items);
            notifyItemRangeInserted(position, items.size());
        }
    }

    public void setItem(final int position, final T item) {
        mData.set(position, item);
        notifyItemChanged(position);
    }

    public List<T> getData() {
        return mData;
    }

    public void setData(final Collection<? extends T> data) {
        mData.clear();
        if (data != null) {
            mData.addAll(data);
        }

        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public T getItem(final int position) {
        return mData.get(position);
    }
}

File: app/src/main/java/siarhei/luskanau/iot/doorbell/ui/base/adapter/BaseRecyclerArrayAdapter.kt
package siarhei.luskanau.iot.doorbell.ui.base.adapter

import android.support.v7.widget.RecyclerView
import java.util.*

abstract class BaseRecyclerArrayAdapter<T, V : RecyclerView.ViewHolder> : BaseRecyclerClickableAdapter<V>() {

    private val items = ArrayList<T>()

    fun addItem(item: T) {
        items.add(item)
        notifyItemInserted(items.size - 1)
    }

    fun addItem(position: Int, item: T) {
        items.add(position, item)
        notifyItemInserted(position)
    }

    fun addItems(items: Collection<T>?) {
        val oldSize = itemCount
        this.items.addAll(items ?: emptyList())
        notifyItemRangeInserted(oldSize, itemCount - oldSize)
    }

    fun addItems(position: Int, items: Collection<T>?) {
        this.items.addAll(position, items ?: emptyList())
        notifyItemRangeInserted(position, items?.size ?: 0)
    }

    fun setItem(position: Int, item: T) {
        items[position] = item
        notifyItemChanged(position)
    }

    fun getItems(): List<T> = items

    fun setItems(items: Collection<T>?) {
        this.items.clear()
        this.items.addAll(items ?: emptyList())
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int = items.size

    fun getItem(position: Int): T = items[position]
}